File: algorithms/algorithms-cplusplus/leetcode/Question_0638_Shopping_Offers.cc
#include <vector>

using namespace std;

class Solution {
 public:
  int shoppingOffers(vector<int>& price, vector<vector<int>>& special, vector<int>& needs) {
    return shopping(price, special, needs);
  }

  int shopping(vector<int>& price, vector<vector<int>>& special, vector<int>& needs) {
    int i = 0, res = dot(needs, price);
    for (vector<int>& s : special) {
      vector<int> clone(needs);
      for (i = 0; i < needs.size(); i++) {
        int diff = clone[i] - s[i];
        if (diff < 0) {
          break;
        }
        clone[i] = diff;
      }
      if (i == needs.size()) {
        res = min(res, s[i] + shopping(price, special, clone));
      }
    }
    return res;
  }

  int dot(vector<int>& a, vector<int>& b) {
    int sum = 0;
    for (int i = 0; i < a.size(); i++) {
      sum += a[i] * b[i];
    }
    return sum;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0638_Shopping_Offers.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0638_Shopping_Offers {

    public int shoppingOffers(List<Integer> price, List<List<Integer>> special, List<Integer> needs) {
        return help(price, special, needs, 0);
    }

    public int help(List<Integer> price, List<List<Integer>> special, List<Integer> needs, int pos) {
        int sum = 0, len = price.size();
        for (int i = 0; i < len; i++) {
            sum += needs.get(i) * price.get(i);
        }
        for (int i = pos; i < special.size(); i++) {
            List<Integer> list = special.get(i);
            List<Integer> temp = new ArrayList<>();
            for (int j = 0; j < needs.size(); j++) {
                if (needs.get(j) < list.get(j)) { // check if the current offer is valid
                    temp = null;
                    break;
                }
                temp.add(needs.get(j) - list.get(j));
            }
            if (temp != null) { // use the current offer and try next
                sum = Math.min(sum, list.get(list.size() - 1) + help(price, special, temp, i));
            }
        }
        return sum;
    }

}
